### PR TITLE
Add support for React Native 0.70

### LINKED ISSUE #3348: Bundled Hermes usage
## Description
`react-native-reanimated` uses `hermes-engine` package in builds on Android. This could potentially break some apps/builds.

### Expected behavior
Use bundled hermes which is located in `node_modules/react-native/sdks`.

### Actual behavior & steps to reproduce
--

## Snack or minimal code example

https://github.com/software-mansion/react-native-reanimated/blob/main/android/build.gradle#L443

## Package versions

| name                                 | version |
| ------------------------------------ | ------- |
| react-native                         |     0.69.1    |
| react-native-reanimated              |     3.0.0-rc.0    |

## Affected platforms

- [x] Android
- [ ] iOS (Probably)
- [ ] Web

### PR DESCRIPTION
**Pull request by kacperkapusciak**
## Description

This PR upgrades FabricExample to use React-Native 0.70.0-rc.2

https://react-native-community.github.io/upgrade-helper/?from=0.69.0&to=0.70.0-rc.2

Closes #3348.

### Changes
* Bumped react-native to 0.70.0-rc.2 in root project, Example and FabricExample apps
* Updated FabricExample app template according to upgrade helper tool
* Partially updated Example app template according to upgrade helper tool (some files differ too much)
* Disabled codegen autolinking on Android in FabricExample app that causes problems
* Changed the way how Reanimated consumes Hermes (see below for details)
* Renamed `resolveBuildType` function to `resolveClientSideBuild`
* Added `resolveBuildType`  that returns `debug` or `release`
* Removed patch-package for react-native with Hermes artifacts cache when building for Android + updated CI scripts
* Added patch-package for gesture-handler and screens that removes explicit Folly version iOS
* Bumped `@babel/core` and `@babel/runtime` in Example app to match the versions in FabricExample
* Updated Reanimated Babel plugin test snapshot and confirmed that `Object.assign` works properly in worklets

### Linking Hermes with Reanimated on Android

Previously, `react-native` declared a dependency on `hermes-engine` npm package. The package contained Hermes header files (i.e. `hermes/hermes.h`) as well as AARs with pre-built Hermes shared library (`libhermes.so`). A mismatch in React Native and Hermes version can result in ABI incompatibility (i.e. unexplainable crashes).

Starting from React Native 0.69, a specific version of Hermes is distributed with React Native (this model is called "Bundled Hermes") to ensure ABI compatibility.

On the new architecture, Hermes is built from sources. React Native downloads Hermes sources from GitHub as tarball, unzips the archive, and compiles Hermes using CMake (see [build.gradle](https://github.com/facebook/react-native/blob/b4f6262bccc7ab96c23f53982701fb0e09752d97/ReactAndroid/hermes-engine/build.gradle#L63-L85)). In Reanimated, we only need to pass a path to Hermes headers (namely `hermes/hermes.h` and `hermes/Public/RuntimeConfig.h`) and the shared library (`libhermes.so`) in `CMakeLists.txt`.

On the old architecture, `libhermes.so` is taken directly from AAR which is embedded in `react-native` npm package (e.g. `node_modules/react-native/android/com/facebook/react/hermes-engine/0.70.0-rc.2/hermes-engine-0.70.0-rc.2-debug.aar`). Hermes sources (and thus headers) are not downloaded by react-native. In Reanimated, we need to download the sources manually (using Gradle tasks copied from `hermes-engine/build.gradle`) as well as unpack `libhermes.so` from AAR in `node_modules/react-native` in order to link it with `libreanimated.so`.

#### References

* https://reactnative.dev/architecture/bundled-hermes
* https://github.com/reactwg/react-native-new-architecture/discussions/4

### CODE REVIEW

**Comment by tomekzaw** [comment]
Hello @kacperkapusciak! Thanks for your contribution.

While the PR seems to work fine on the CI, it fails when compiling C++ sources locally:

Here are the logs from CMake:

<details>

```
> Task :app:buildNdkBuildDebug[arm64-v8a][fabricexample_appmodules] FAILED
C/C++: In file included from /Users/tomekzaw/RNOS/rn070/react-native-reanimated/FabricExample/android/app/src/main/jni/MainApplicationTurboModuleManagerDelegate.cpp:1:
C/C++: /Users/tomekzaw/RNOS/rn070/react-native-reanimated/FabricExample/android/app/src/main/jni/MainApplicationTurboModuleManagerDelegate.h:25:53: error: non-virtual member function marked 'override' hides virtual member functions
C/C++:       const std::shared_ptr<CallInvoker> jsInvoker) override;
C/C++:                                                     ^
C/C++: /Users/tomekzaw/RNOS/rn070/react-native-reanimated/FabricExample/android/../node_modules/react-native/ReactAndroid/src/main/java/com/facebook/react/turbomodule/core/jni/ReactCommon/TurboModuleManagerDelegate.h:25:40: note: hidden overloaded virtual function 'facebook::react::TurboModuleManagerDelegate::getTurboModule' declared here: type mismatch at 2nd parameter ('const JavaTurboModule::InitParams &' vs 'const std::shared_ptr<CallInvoker>')
C/C++:   virtual std::shared_ptr<TurboModule> getTurboModule(
C/C++:                                        ^
C/C++: /Users/tomekzaw/RNOS/rn070/react-native-reanimated/FabricExample/android/../node_modules/react-native/ReactAndroid/src/main/java/com/facebook/react/turbomodule/core/jni/ReactCommon/TurboModuleManagerDelegate.h:28:40: note: hidden overloaded virtual function 'facebook::react::TurboModuleManagerDelegate::getTurboModule' declared here: type mismatch at 2nd parameter ('const std::shared_ptr<CallInvoker> &' vs 'const std::shared_ptr<CallInvoker>')
C/C++:   virtual std::shared_ptr<TurboModule> getTurboModule(
C/C++:                                        ^
C/C++: /Users/tomekzaw/RNOS/rn070/react-native-reanimated/FabricExample/android/app/src/main/jni/MainApplicationTurboModuleManagerDelegate.cpp:24:44: error: out-of-line definition of 'getTurboModule' does not match any declaration in 'facebook::react::MainApplicationTurboModuleManagerDelegate'
C/C++: MainApplicationTurboModuleManagerDelegate::getTurboModule(
C/C++:                                            ^~~~~~~~~~~~~~
C/C++: /Users/tomekzaw/RNOS/rn070/react-native-reanimated/FabricExample/android/app/src/main/jni/MainApplicationTurboModuleManagerDelegate.cpp:38:49: error: out-of-line definition of 'canCreateTurboModule' does not match any declaration in 'facebook::react::MainApplicationTurboModuleManagerDelegate'
C/C++: bool MainApplicationTurboModuleManagerDelegate::canCreateTurboModule(
C/C++:                                                 ^~~~~~~~~~~~~~~~~~~~
C/C++: /Users/tomekzaw/RNOS/rn070/react-native-reanimated/FabricExample/android/app/src/main/jni/MainApplicationTurboModuleManagerDelegate.h:34:35: note: type of 1st parameter of member declaration does not match definition ('const std::string &' (aka 'const basic_string<char, char_traits<char>, allocator<char>> &') vs 'std::string &' (aka 'basic_string<char, char_traits<char>, allocator<char>> &'))
C/C++:   bool canCreateTurboModule(const std::string &name);
C/C++:                                   ^
C/C++: /Users/tomekzaw/RNOS/rn070/react-native-reanimated/FabricExample/android/app/src/main/jni/MainApplicationTurboModuleManagerDelegate.cpp:40:31: error: reference to type 'const JavaTurboModule::InitParams' could not bind to an rvalue of type 'nullptr_t'
C/C++:   return getTurboModule(name, nullptr) != nullptr ||
C/C++:                               ^~~~~~~
C/C++: /Users/tomekzaw/RNOS/rn070/react-native-reanimated/FabricExample/android/app/src/main/jni/MainApplicationTurboModuleManagerDelegate.cpp:34:40: note: passing argument to parameter 'params' here
C/C++:     const JavaTurboModule::InitParams &params) {
C/C++:                                        ^
C/C++: In file included from /Users/tomekzaw/RNOS/rn070/react-native-reanimated/FabricExample/android/app/src/main/jni/MainApplicationTurboModuleManagerDelegate.cpp:1:
C/C++: In file included from /Users/tomekzaw/RNOS/rn070/react-native-reanimated/FabricExample/android/app/src/main/jni/MainApplicationTurboModuleManagerDelegate.h:4:
C/C++: In file included from /Users/tomekzaw/RNOS/rn070/react-native-reanimated/FabricExample/android/../node_modules/react-native/ReactAndroid/src/main/java/com/facebook/react/turbomodule/core/jni/ReactCommon/TurboModuleManagerDelegate.h:11:
C/C++: In file included from /Users/tomekzaw/RNOS/rn070/react-native-reanimated/FabricExample/android/../node_modules/react-native/ReactAndroid/../ReactCommon/react/nativemodule/core/platform/android/ReactCommon/JavaTurboModule.h:18:
C/C++: In file included from /Users/tomekzaw/RNOS/rn070/react-native-reanimated/FabricExample/android/../node_modules/react-native/ReactAndroid/src/main/jni/react/jni/JCallback.h:12:
C/C++: In file included from /Users/tomekzaw/RNOS/rn070/react-native-reanimated/FabricExample/android/../node_modules/react-native/ReactAndroid/src/main/jni/first-party/fbjni/headers/fbjni/fbjni.h:30:
C/C++: /Users/tomekzaw/RNOS/rn070/react-native-reanimated/FabricExample/android/../node_modules/react-native/ReactAndroid/src/main/jni/first-party/fbjni/headers/fbjni/detail/Hybrid.h:206:50: error: allocating an object of abstract class type 'facebook::react::MainApplicationTurboModuleManagerDelegate'
C/C++:     return makeHybridData(std::unique_ptr<T>(new T(std::forward<Args>(args)...)));
C/C++:                                                  ^
C/C++: /Users/tomekzaw/RNOS/rn070/react-native-reanimated/FabricExample/android/app/src/main/jni/MainApplicationTurboModuleManagerDelegate.cpp:10:10: note: in instantiation of function template specialization 'facebook::jni::HybridClass<facebook::react::MainApplicationTurboModuleManagerDelegate, facebook::react::TurboModuleManagerDelegate>::makeCxxInstance<>' requested here
C/C++:   return makeCxxInstance();
C/C++:          ^
C/C++: /Users/tomekzaw/RNOS/rn070/react-native-reanimated/FabricExample/android/../node_modules/react-native/ReactAndroid/src/main/java/com/facebook/react/turbomodule/core/jni/ReactCommon/TurboModuleManagerDelegate.h:28:40: note: unimplemented pure virtual method 'getTurboModule' in 'MainApplicationTurboModuleManagerDelegate'
C/C++:   virtual std::shared_ptr<TurboModule> getTurboModule(
C/C++:                                        ^
C/C++: 5 errors generated.
C/C++: make: *** [/Users/tomekzaw/Library/Android/sdk/ndk/24.0.8215888/build/core/build-binary.mk:424: /Users/tomekzaw/RNOS/rn070/react-native-reanimated/FabricExample/android/app/build/intermediates/cxx/Debug/i3gi6c6m/obj/local/arm64-v8a/objs-debug/fabricexample_appmodules//Users/tomekzaw/RNOS/rn070/react-native-reanimated/FabricExample/android/app/src/main/jni/MainApplicationTurboModuleManagerDelegate.o] Error 1
```

</details>

**Comment by kacperkapusciak** [comment]
Hello @tomekzaw 👋 
Whoops! It should work on the newest main 🙊 Try to merge the newest main branch to this branch 🧇 🍖 

Cheers

**Comment by tomekzaw** [comment]
😢 

Zrzut ekranu 2022-08-1 o 13 08 07

**Comment by tomekzaw** [comment]
TODO:
* check if `Object.assign` is whitelisted in worklets
* change Hermes AAR path in patch-packages for optimizing CI speed
* don't copy source code of `unzipHermes` task

File: android/build.gradle [hunk 1]
```diff
@@ -438,24 +452,95 @@ task prepareGlog(dependsOn: dependenciesPath ? [] : [downloadGlog], type: Copy)
...
+
+    def reactNativeRootDir = resolveReactNativeDirectory()
+    def customDownloadDir = System.getenv("REACT_NATIVE_DOWNLOADS_DIR")
+    def downloadDir = customDownloadDir ? new File(customDownloadDir) : new File(reactNativeRootDir, "sdks/download")
+
+    // By default we are going to download and unzip hermes inside the /sdks/hermes folder
+    // but you can provide an override for where the hermes source code is located.
+    def overrideHermesDir = System.getenv("REACT_NATIVE_OVERRIDE_HERMES_DIR") != null
+    def hermesDir = System.getenv("REACT_NATIVE_OVERRIDE_HERMES_DIR") ?: new File(reactNativeRootDir, "sdks/hermes")
+    def hermesBuildDir = new File("$buildDir/hermes")
```
Comment (piaskowyk) [inline android/build.gradle:476]: The variables `overrideHermesDir` and `hermesBuildDir` seems to be unused.

Comment (tomekzaw) [reply android/build.gradle]: Removed in bd640afcea13d51893e8a970ace60f113af8693f

**Review by piaskowyk — APPROVED** [review]
Well done!

**Pull request merged**